### ISSUE TITLE
Feature Request - Allow remote configuration of plugin via polling

### ISSUE THREAD

**Issue by bryanrideshark** [transistorsoft/cordova-background-geolocation-lt#927]
# **tl;dr**
I'd like to be able to specify a URL, which this plugin would poll at a specified interval to retrieve the configuration. Using this method, I'd be able to change the settings of when locations should be tracked etc. on our server, and the app would be able to cancel scheduled tracking / add new scheduled tracking times, based on the servers' response.

## Motivations:
Our application has both a web-app, and a dedicated mobile app (cordova). Our app is used to help coordinate meetings in the physical world between two or more users. Users can add new meetings, cancel meetings, etc. from either the web app, or the mobile app. Of course, this allows for the configuration of the Background Geolocation plugin to become out of sync.

A common error we're encountering is the case where a user cancels a meeting on the web-app, and the mobile app on their phone is still tracking their location. The only way for them to stop this activity is to open the app, upon which our Javascript makes calls to the server, realizes the scheduled location tracking times have stopped, and updates the plugin with the new configuration.

## Desired features:
The only thing we'd need, is the ability to specify which URL should be polled to update the configuration, as well as the options for how polling should take place.

Example schema (Typescript notation)
````ts
export interface BackgroundGeolocationPluginPollingConfiguration {
  pollingUrl: string;
  // Some sort of way we can schedule the behavior, 
  // example values could be "every 1 s", "every 15 min", "every 1 day", something like that.
  pollingInterval: string;

  // Specified times when polling should take place. (used in addition to or instead of pollingInterval).
  // For our use case, we would likely want to specify polling to take place leading up to a meeting, for example, we'd poll an hour, half an hour, fifteen minutes, five minutes, and one minute before the meeting took place. In this way, hopefully we'd catch the case where the meeting was cancelled, in which case we'd stop tracking the users' location
  pollingTimes: string[];
}

````

Ideally, the polled configuration would allow us to configure every aspect of this plugin, including the configuration for polling itself. In this way, we'd be able to avoid having to write extensive Javascript merely to synchronize the state of the configuration on the device, with the desired state we have on the server.

Obviously, it would be great if we could also update the state of this plugin by pushing from the server, but I imagine that would be more difficult to configure. Polling (should be?) simple to implement, and if you could add it to the capabilities of the plugin we'd greatly appreciate it.

**Comment by christocracy** [comment]
Polling is a really bad idea on mobile, it consumes a lot of energy.

Why can’t you have your server send “silent” push notifications to the device(s) with a JSON payload, whenever you desire a change to some particular device’s config?

**Comment by christocracy** [comment]
And why can’t you use background-fetch for your periodic tasks?  That’s what it’s made for.

**Comment by bryanrideshark** [comment]
Hi Chris, thanks for the quick response.

I would love to handle this via push notifications - we already use FCM (firebase cloud messaging?) push notifications in our app. But as far as I know, there's no way to change the configuration of this plugin upon receiving such a notification.

Is there a way to do that upon receiving a push notification from Firebase, or some other push notification service, with a Cordova app? If there is, we'd gladly do so as it would be a much better solution than polling.

**Comment by bryanrideshark** [comment]
The fact that background-fetch is a thing must have totally gone over my head. That should be enough to get started with something, thank you.

Still, if there is a way to update the configuration of this plugin with a push notification, that would be ideal.

**Comment by christocracy** [comment]
>But as far as I know, there's no way to change the configuration of this plugin upon receiving such a notification.

The plugin doesn't need to be involved with Push.

```typescript
// It's your responsibility to set up this callback and define the schema of the data
onPushNotification(notification) { 
  // pseudo code.
  if (notification.data) {
    let data = JSON.parse(notification.data);
    if (data.backgroundGeolocation) {
      BackgroundGeolocation.setConfig(data.backgroundGeolocation);
    }
  }
}
```

And for the Android case where your app is terminated, you must implement [Android Headless Mode](https://github.com/transistorsoft/cordova-background-geolocation/wiki/Android-Headless-Mode).

**Comment by christocracy** [comment]
BackgroundGeolocation was originally designed for a disaster-response app, tracking first-responders in hurricanes, earthquakes.

This app allowed one to enter a secured area guarded by police / military.  When one wanted to enter the danger zone, the guard scanned a QR code on the workers phone.  The guard's device sent an HTTP request to server.  If the worker was authorized to enter, the server sent a Push notification to the their device, the payload parsed (as above), and `BackgroundGeolocation.start` was executed, tracking them for as long as they were in the secured area.

On leaving the secured area, the guard scanned them out, a push was sent to the device once again, the payload parsed and `BackgroundGeolocation.stop`.

**Comment by bryanrideshark** [comment]
Right, so in your scenario it was highly likely that the app was already running in the foreground, in which case it would be simple to change the configuration of BackgroundGeolocation.

I'll have to see what can be done about our existing push notification infrastructure, and whether or not we'd be able to use it to change the configuration of BackgroundGeolocation. Thank you for your time!

**Comment by christocracy** [comment]
This obsolete plugin of mine [cordova-plugin-background-notification](https://github.com/christocracy/cordova-plugin-background-notification) was designed at the same time I created `background-geolocation` and `background-fetch` to receive notification in the background from Urban Airship.

Yes, you can receive push notification in the background / terminated states.

On iOS, that's what the *Background Mode* `[x] Remote notifications` is for.
![](https://dl.dropbox.com/s/lib4vun5jdmyc8l/Screenshot%202019-05-08%2017.36.09.png?dl=1)

**Comment by christocracy** [comment]
> it was highly likely that the app was already running in the foreground

Push notifications can take some seconds.  It couldn't be guaranteed the worker still had his phone open.  That's why it was required to receive push in background.

Also, there were cases where entering / exiting a geofence would cause the server to send a silent push to start/stop/reconfigure tracking.

**Comment by bryanrideshark** [comment]
I believe our app currently uses https://github.com/phonegap/phonegap-plugin-push. It's unfortunate that your background push plugin isn't currently supported, as from what I've been able to gather, running handlers in the background is not possible with our current plugin.

**Comment by christocracy** [comment]
I *instructed* `phonegap-plugin-push` [how to handle background push notifications](https://github.com/phonegap/phonegap-plugin-push/issues/93) in 2015.

`phonegap-plugin-push` has been able to handle push in background for *years*.
